### ISSUE TITLE
Question about all_masks shape in the ppo_trainer.py file

### ISSUE THREAD

**Issue by liwenju0** [huggingface/trl#646]
in ppo_trainer.py,  i  am confused with the all_masks shape。all_masks comes from masks, masks have already cut tail element as the attention_mask, but the all_masks cut tail element again, i don't  know why this is right? i think cut one time is enough. thank you!
<img width="688" alt="image" src="https://github.com/huggingface/trl/assets/16049564/6d721334-9af4-41a3-83ea-8a6d273f8dba">
why here cut mask two times？

**Comment by lvwerra** [comment]
Ok, so the first time we don't actually cut it but just fill only the all values until before the last one. That's because the attention mask is aligned with the inputs and the mask should be aligned with the outputs. They are offset by 1 as the model predicts the next token.

Finally at the end we cutoff the last element as this corresponds to the next element that would have been predicted (but wasn't (see how we slice the `logits` it `logprobs_from_logits`). 

Does that make sense?

**Comment by liwenju0** [comment]
ok, thank you very much! very helpful!
